File: TP01/TP01/palindromo.cpp
/*--------------------TP01---------------------
Aluna: Ana Laura Fernandes de Oliveira
Matricula: 680715
Disciplina: Algoritmos e Estrutura de dados II
Professor: Max do Val Machado
----------------------------------------------- */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

bool isPalindromo(char* frase)
{
    //declaracao de variaveis
    bool resposta = false;
    int j = 0;
    int tamanho = 0;

    if(frase != NULL)
    {
        //obter o tamanho da string
        tamanho = strlen(frase);

        //verificar se palavra lida e um palindromo
        for(int i = tamanho - 2; i >= 0; i--) // o -2 e para eliminar o \n lido pelo fgets e para chegar na ultima posicao valida
        {
            if(frase[i] != frase[j]) //ao encontrar um caracter diferente, ja atribui o valor de falso e sai do loop
            {
                resposta = false;
                i = -1;
            }
            else
            {
                resposta = true;
            }
            j++;
        }
    }
    
    return resposta;
}
bool isFim(char* frase)
{
    return(frase[0] == 'F' && frase[1] == 'I' && frase[2] == 'M');
}
int main(int argc, char* argv[])
{
    //declaracao de variaveis
    char* entrada;
    int numEntradas = 0;
    bool resposta = false;
    
    //alocar memoria dinamicamente para receber as entradas
    entrada = (char*) malloc (1000 * sizeof(char));
    
    //receber entradas
    fgets(entrada, 1000, stdin);

    while(!(isFim(entrada)))
    {
        //testar se 'string' lida e um palindromo
        resposta = isPalindromo(entrada);
        
        if(resposta)
        {
            printf("%s\n", "SIM");
        }
        else
        {
            printf("%s\n", "NAO");
        }
        fgets ( entrada, 1000, stdin );
    }
    
    //liberar memoria alocada
    free(entrada);

    //fim do programa
    fflush(stdin);
    return 0;
}

/*----------------Testes-----------------
Input                                     /Output
ana                                       SIM
Algoritmos e Estrutura de Dados II        NAO
teste do algoritmo                        NAO
ciencia da computacao                     NAO
abacaxi ixacaba                           SIM
*/ 

File: TP01/TP01/palindromo.java
/*----------------------TP01---------------------
 *Aluna: Ana Laura Fernandes de Oliveira
 *Disciplina: Algoritmos e Estrutura de Dados II
 *Professor: Max do Val Machado
------------------------------------------------ */

public class palindromo
{
    public static boolean isPalindromo(String palavra)
    {
	//declaracao de variaveis    
        String palindromo = palavra;
        int j = 0;
        boolean resposta = false;
	
	//passos de verificacao
        for(int i = palavra.length() - 1; i >= 0 ; i--)
        {
            //verificando se a palavra, se lida da forma invertida eh a mesma que lida da esquerda pra direita
            if(palindromo.charAt(j) == palavra.charAt(i))
            {
                resposta = true;
            }
            else
            {
                resposta = false;
                i = -1;
            }
            j++;
        }

        return(resposta);
    }//end isPalindromo
    
    public static boolean isFim (String entrada)
    {
        return(entrada.charAt(0) == 'F' && entrada.charAt(1) == 'I' && entrada.charAt(2) == 'M');
    }
    public static void main(String[] args)
    {
        //declarar variaveis
        String[] palavras = new String[1000];
	int i = 0;
        
        
	    //receber entradas
            palavras[i] = MyIO.readLine();
	while(isFim(palavras[i]) == false)
	{
	    
	    //testar mediante uso de funcao se a string recebida e um palindromo e imprimir na tela o resultado
            if(isPalindromo(palavras[i]))
            {
                MyIO.println ("SIM");
            }
            else
            {
                MyIO.println("NAO");
            }
	    i++;
	    palavras[i] = MyIO.readLine();
        }
    }
}
/*---------------------Testes----------------------
 * ENTRADA                   SAIDA
 * ana                       SIM
 * hannah                    SIM
 * banana                    NAO
 * ciencia da comp           NAO
 * ama                       SIM
 * abacaxi                   NAO
 * ferias                    NAO
 * ------------------------------------------------*/
